### ISSUE TITLE
Expression: Support pickling 

### ISSUE THREAD

**Issue by nepfaff** [RobotLocomotion/drake#20673]
**Is your feature request related to a problem? Please describe.**
I'm computing symbolic expressions that stay constant for my system and computing them takes a very long time. I want to compute them once, store them on disk, and then load them whenever I need them.

**Describe the solution you'd like**
Pickle support for symbolic `Expression`s. Adding pickle support for `Expression` was previously mentioned [here](https://stackoverflow.com/questions/58982413/serializing-drake-objects-with-python).

**Describe alternatives you've considered**
Converting the expressions to `sympy` (using `to_sympy`) and pickling these. This solution works. However, I wanted to create this issue anyway, as going through `sympy` for pickle support seems a bit odd, and spending offline time computing equations seems like a decently common use case. The `sympy` option makes this a low priority.

**Comment by EricCousineau-TRI** [comment]
Assigning myself nominally per component leads.
\cc @jwnimmer-tri @hongkai-dai

**Comment by jwnimmer-tri** [comment]
FYI When we circle back here -- the implementation tactic should be to reuse the `Unapply` logic from `symbolic_py_unapply.h`, but for pickle we'd want to use `(expression_kind, (args...))` instead of `(ctor_func, (args..))`.  Then when unpickling we'd look up the `ctor_func` based on the `expression_kind`.

**Comment by j4yyousi** [comment]
Hi, I realized that to pickle `Expression`s we need to pickle `Variable`s, since `Expression`s can contain `Variable`s. Is it a requirement that `Variable`s must have the same id before and after pickling? Or we can store only `Variable`s’ name and type, then when unpickling reconstruct the `Variable` objects with different ids?

**Comment by jwnimmer-tri** [comment]
The IDs don't need to be the same, but they do need to be a bijection between the old IDs and new IDs.

Variable names are not unique, only IDs are, so pickling just the name is insufficient.

**Comment by j4yyousi** [comment]
Why do we need to maintain a bijection between the old and new IDs? 
My understanding is that an ID contains the variable’s type in its most significant byte and a unique integer from a global counter in the remaining 7 bytes. So storing the variable’s name and type would be sufficient, since the global counter will assign a new integer to the reconstructed variable.  Also, my assumption here is that program A will pickle variables while program B will load them, where each program has its own global counter. Thus I can’t see why it would be useful to maintain a bijection between different program instances.

**Comment by jwnimmer-tri** [comment]
> So storing the variable’s name and type would be sufficient, since the global counter will assign a new integer to the reconstructed variable.

No.  Names are not unique, and do not uniquely identify a variable:

```py
>>> from copy import copy
>>> from pydrake.all import Variable
>>> x1 = Variable("x")
>>> x2 = Variable("x")
>>> e = (2*x1+3*x2)*(4*copy(x1)+5*copy(x2))
>>> e
<Expression "((2 * x + 3 * x) * (4 * x + 5 * x))">
>>> e.Evaluate({x1: 1, x2: 10})
1728.0
```

If I pickle and then unpickle `e`, I need it to come out the other side exactly the same.  If I evaluate it with x1=1 and x2=10 after unpickling, I need to get `1728` back again (32 * 54).

The uniquely identifying property of a variable is its ID (`x1.get_id()`), not its name, and not its python object identity (`id(x1)`).  It necessarily requires special handling during pickling.

**Comment by j4yyousi** [comment]
Thanks for clarifying! I got confused by assuming you meant a physical bijection: [a data structure like memo](https://github.com/RobotLocomotion/drake/blob/77d32a47b0cbc137485e1d2b7902b5b05c905c3e/bindings/pydrake/symbolic/_symbolic_extra.py#L50). From the example I understood that you meant a logical bijection: for every unique variable encountered  during pickling, a unique variable must be created when unpickling.

**Comment by j4yyousi** [comment]
Hi @jwnimmer-tri, I’d appreciate your input on the following pickle/unpickle scenario:

Suppose a user pickles an expression using `store.py` and loads it using `load.py`
```python
#store.py
import pydrake.symbolic as sym
import pickle
x = sym.Variable("x"); y = sym.Variable("y")
z = sym.Variable("z"); w = sym.Variable("w")
a = sym.Variable("a")
e = sym.pow(a*(x+y), w+z)
print(e.GetVariables()) # {x, y, z, w, a} same as creation order, 
# since GetVariables returns an ordered set in ascending order 
# by comparing variable ID.
with open("expr.pkl", "wb") as f:
    pickle.dump(e, f)
```

```python
#load.py
from pydrake.symbolic import *
import pickle
with open("expr.pkl", "rb") as f:
    e = pickle.load(f)
vars = e.GetVariables() # my WIP returns the following order {a, x, y, z, w}.
# During unpickling, variables can get reconstructed at a different order than
# store.py, but still deterministic.
env = {x:1, y:2, z:3, w:4, a:5} # error, since x, y, z, w, a are not defined.
e.Evaluate(env)
```
After unpickling an expression, how is the user supposed to define an environment, without having the variables `{x, y, z, w, a}` explicitly defined as in `store.py`?
The solution I could think of is the following, which is very manual:
```python
vars = e.GetVariables() # suppose it returns {a, x, y, z, w}. Order is 
# deterministic, but unkown and one needs to check print(vars) once to know it.
vars_list = list(vars) #[a, x, y, z, w]
a = vars_list[0]
x = vars_list[1]
y = vars_list[2]
z = vars_list[3]
w = vars_list[4]
env = {x:1, y:2, z:3, w:4, a:5}  #Now works
e.Evaluate(env)
```

**Comment by j4yyousi** [comment]
Hi, I'm also curious to know why `Variable` [equality is checked by ID](https://github.com/RobotLocomotion/drake/blob/fbf42ba67d0e54a98a47a39bd5eb1d23fa476c77/common/symbolic/expression/variable.h#L86)? Since [in Sympy, two variables are equal](https://github.com/sympy/sympy/blob/a17c0ce85a7968f121985d65cb23ae0b9e1b9ecc/sympy/core/basic.py#L526) if their class attributes are equal (it assumes unique names), there is no ID system. Which allows the following pickle/unpickle to easily work in Sympy:
```python
#store_sympy.py
from sympy import symbols, sqrt, pi, evaluate
import pickle
with evaluate(False):
    x_sym, y_sym = symbols('x y')
    expr_to_pickle = sqrt(x_sym**2 + y_sym**2 + pi)
with open("expr.pkl", "wb") as f:
    pickle.dump(expr_to_pickle, f)
```

```python
#load_sympy.py
import sympy
import pickle

z_sym, w_sym = sympy.symbols('z z')
print(z_sym == w_sym) #output: True

with open("expr.pkl", "rb") as f:
    unpickled_expr = pickle.load(f)
x_sym, y_sym = sympy.symbols('x y')
print(x_sym == y_sym) #output: False
values_dict = {x_sym: 3, y_sym: 4} #works
print(unpickled_expr.evalf(subs=values_dict))
```

**Comment by jwnimmer-tri** [comment]
> ... how is the user supposed to define an environment ...

The easiest way would be instead of doing `pickle.dump(e, f)`, users would do something like `pickle.dump((e, vars), f)` and then `e, vars = pickle.load(f)`, where vars is either a list `[a, x, y, z, w]` or a dict `("a": a, "x": x, etc.)`.

Of course they also could only pickle `e` and then use `e.GetVariables()` and then make up a dict based on the name, if they know apriori that they are supposed to be unique:

```py
vars = dict([
    (item.get_name(), item)
    for item in e.GetVariables()
])
```

> Hi, I'm also curious to know why Variable equality is checked by ID?

That's just the way the code was written many years ago.

**Comment by j4yyousi** [comment]
Hi, I'm currently implementing the solution in Python utilizing the `Unapply` function. Is there a preference on which side to code? i.e. should the solution be in C++ and bound to Python through pybind?

**Comment by jwnimmer-tri** [comment]
Generally I'd go with whatever leads to the simplest code.  That probably means implementing in Python, but if parts are easier in C++, that's fine too.

**Comment by j4yyousi** [comment]
Regarding 
> ... e, vars = pickle.load(f), where vars ...

pickle returns the same object being pickled: `e = pickle.loads(pickle.dumps(e))` . Thus users will have to opt for 
> ... only pickle e and then use e.GetVariables() and then make up a dict based on the name, if they know apriori that they are supposed to be unique ...
